src/embed.ts

```diff
@@ -94,38 +94,33 @@ const VimeoApiResponse = t.type({
 })
 
 async function getVimeoThumbnail(url: URL) {
-  // exmaple url: https://player.vimeo.com/video/${id}?autoplay=1
-
-  const videoId = url.pathname.replace('/video/', '')
-
-  if (!RegExp('[0-9]+').test(videoId)) {
-    return getPlaceholder()
-  }
-  const apiResponse = await fetch(
-    'https://vimeo.com/api/oembed.json?url=' +
-      encodeURIComponent(`https://vimeo.com/${videoId}`)
-  )
-  if (apiResponse.status !== 200) return getPlaceholder()
-
-  try {
-    const thumbnailUrl = pipeable.pipe(
-      VimeoApiResponse.decode(await apiResponse.json()),
-      O.fromEither,
-      O.map((response) => response.thumbnail_url.replace(/_[0-9|x]+/, '')),
-      O.chain((url) => O.tryCatch(() => new Url(url))),
-      O.chain(O.fromPredicate((url) => url.domain === 'vimeocdn.com'))
-    )
-
-    if (O.isNone(thumbnailUrl)) return getPlaceholder()
-
-    const imgRes = await fetch(thumbnailUrl.value.href)
-
-    return imgRes.status === 200 ? imgRes : getPlaceholder()
-  } catch (e) {
-    // Error with parsing the json request
-  }
-
-  return getPlaceholder()
+  const error = () => 'error' as unknown
+
+  return pipeable.pipe(
+    TE.of(url.pathname.replace('/video/', '')),
+    TE.chain(TE.fromPredicate((videoId) => /[0-9]+/.test(videoId), error)),
+    TE.chain((videoId) =>
+      TE.fromTask(() =>
+        fetch(
+          'https://vimeo.com/api/oembed.json?url=' +
+            encodeURIComponent(`https://vimeo.com/${videoId}`)
+        )
+      )
+    ),
+    TE.chain(
+      TE.fromPredicate((apiResponse) => apiResponse.status === 200, error)
+    ),
+    TE.chain((apiResponse) => TE.tryCatch(() => apiResponse.json(), error)),
+    TE.chain((data) =>
+      TE.fromEither(E.mapLeft(error)(VimeoApiResponse.decode(data)))
+    ),
+    TE.map((data) => data.thumbnail_url.replace(/_[0-9|x]+/, '')),
+    TE.chain((url) => TE.fromEither(E.tryCatch(() => new Url(url), error))),
+    TE.chain(TE.fromPredicate((url) => url.domain === 'vimeocdn.com', error)),
+    TE.chain((url) => TE.fromTask(() => fetch(url.href))),
+    TE.chain(TE.fromPredicate((response) => response.status === 200, error)),
+    TE.getOrElse(() => () => Promise.resolve(getPlaceholder()))
+  )()
```

Comment:
It's probably mostly because I'm not used to it:
to be honest for me this syntax is not readable at all and the upside is not immediately obvious.

Reply:
Yeah again, I'm not sure we want to turn typescript into rust here.
Maybe it's a topic for our dev meet.
Even though it's probably safer, for me this code is actually worse overall.

Reply:
I am fine with using the old syntax (which is more verbose but better to read). Let's discuss it at monday with the other developers.